File: backend/app/schemas/comparisons.py
# =============================================================================
# File: comparisons.py
# Module/Service: Comparison Service (FR8)
# Layer: Schema
# Purpose: Pydantic request/response models for Comparisons API (OpenAPI).
# Responsibilities:
#   - ComparisonCreateRequest (document_ids ≥2, optional focus)
#   - ComparisonResponse (public contract including async status)
# Dependencies:
#   - Pydantic, app.models.enums
# Public Exports:
#   - ComparisonCreateRequest, ComparisonResultPayload, ComparisonResponse
# Database/Table: comparisons (read mapping only)
# Related Modules: app.api.comparisons, Enterprise_notebooklm_openapi.yaml
# Important Notes:
#   - status enables FE polling (Summary/Extraction convention); result=null
#     while processing/failed.
#   - review / comments are optional CMP-20/22 metadata on the comparison row,
#     never inside result. Analysis fields stay immutable.
#   - audit is served on a dedicated GET; it is not part of ComparisonResponse.
# =============================================================================

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.enums import ComparisonStatus

ComparisonReviewStatus = Literal[
    "OPEN",
    "REVIEWED",
    "NEEDS_ATTENTION",
    "ACKNOWLEDGED",
]

ComparisonCommentTarget = Literal[
    "CLAUSE",
    "FINDING",
    "EXACT_DIFFERENCE",
    "EVIDENCE",
]


class ComparisonCreateRequest(BaseModel):
    document_ids: list[uuid.UUID] = Field(min_length=2)
    focus: str | None = None

    @field_validator("document_ids")
    @classmethod
    def _at_least_two(cls, value: list[uuid.UUID]) -> list[uuid.UUID]:
        if len(value) < 2:
            raise ValueError("document_ids must contain at least 2 items")
        return value


class ComparisonReviewDecision(BaseModel):
    status: ComparisonReviewStatus
    reviewer_id: uuid.UUID | None = None
    reviewer_name: str | None = None
    reviewed_at: datetime | None = None


class ComparisonReviewUpdateRequest(BaseModel):
    clause_id: str = Field(min_length=1, max_length=128)
    status: ComparisonReviewStatus

    @field_validator("clause_id")
    @classmethod
    def _clause_id_not_blank(cls, value: str) -> str:
        text = value.strip()
        if not text:
            raise ValueError("clause_id is required")
        return text


class ComparisonCommentCreateRequest(BaseModel):
    clause_id: str = Field(min_length=1, max_length=128)
    body: str = Field(min_length=1, max_length=4000)
    target_type: ComparisonCommentTarget = "CLAUSE"
    target_id: str | None = Field(default=None, max_length=128)

    @field_validator("clause_id")
    @classmethod
    def _clause_id_not_blank(cls, value: str) -> str:
        text = value.strip()
        if not text:
            raise ValueError("clause_id is required")
        return text

    @field_validator("body")
    @classmethod
    def _body_not_blank(cls, value: str) -> str:
        text = value.strip()
        if not text:
            raise ValueError("body is required")
        return text

    @field_validator("target_id")
    @classmethod
    def _target_id_trim(cls, value: str | None) -> str | None:
        if value is None:
            return None
        text = value.strip()
        return text or None


class ComparisonCommentUpdateRequest(BaseModel):
    body: str = Field(min_length=1, max_length=4000)

    @field_validator("body")
    @classmethod
    def _body_not_blank(cls, value: str) -> str:
        text = value.strip()
        if not text:
            raise ValueError("body is required")
        return text


class ComparisonComment(BaseModel):
    id: str
    clause_id: str
    target_type: Literal["CLAUSE", "EXACT_DIFFERENCE", "EVIDENCE"]
    target_id: str | None = None
    body: str
    author_id: uuid.UUID | None = None
    author_name: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


ComparisonAuditActionName = Literal[
    "CLAUSE_OPENED",
    "REVIEW_STATUS_CHANGED",
    "COMMENT_ADDED",
    "COMMENT_EDITED",
    "COMMENT_DELETED",
]


class ComparisonAuditCreateRequest(BaseModel):
    action: Literal["CLAUSE_OPENED"]
    clause_id: str = Field(min_length=1, max_length=128)

    @field_validator("clause_id")
    @classmethod
    def _clause_id_not_blank(cls, value: str) -> str:
        text = value.strip()
        if not text:
            raise ValueError("clause_id is required")
        return text


class ComparisonAuditEvent(BaseModel):
    id: str
    action: ComparisonAuditActionName
    clause_id: str | None = None
    actor_id: uuid.UUID | None = None
    actor_name: str | None = None
    occurred_at: datetime
    before: dict[str, Any] | None = None
    after: dict[str, Any] | None = None
    target_type: str | None = None
    target_id: str | None = None
    comment_id: str | None = None


class ComparisonAuditTrailResponse(BaseModel):
    events: list[ComparisonAuditEvent] = Field(default_factory=list)


class ComparisonResultPayload(BaseModel):
    similarities: list[str] = Field(default_factory=list)
    differences: list[str] = Field(default_factory=list)
    contract_comparison: dict[str, Any] | None = None


class ComparisonResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: uuid.UUID
    document_ids: list[uuid.UUID]
    status: ComparisonStatus
    result: ComparisonResultPayload | None = None
    review: dict[str, ComparisonReviewDecision] = Field(default_factory=dict)
    comments: list[ComparisonComment] = Field(default_factory=list)
    created_at: datetime

File: frontend/types/comparisons.ts
/**
 * =============================================================================
 * File: comparisons.ts
 * Module/Service: Comparison Service (Web App)
 * Layer: UI
 * Purpose: TypeScript types for Comparisons API matching OpenAPI (FR8 / UC7).
 * Responsibilities:
 *   - Align Comparison / status / result with backend schema
 *   - Optional contract_comparison (CMP-15/16) for clause-level summary UI
 *   - CMP-23 audit events are fetched separately from Comparison payload
 * Dependencies:
 *   - docs/Enterprise_notebooklm_openapi.yaml Comparison schema
 * Public Exports:
 *   - ComparisonStatus, ComparisonResult, Comparison, ComparisonCreateRequest
 *   - ContractComparisonReport and nested clause/evidence types
 * Database/Table: comparisons, comparison_documents
 * Related Modules: lib/comparisons.api, features/comparisons/*
 * Important Notes: result is null while processing/failed; object when completed.
 *   contract_comparison is omitted/null when clause enrichment is unavailable.
 * =============================================================================
 */

export type ComparisonStatus = "processing" | "completed" | "failed";

export type ClauseComparisonStatus =
  | "UNCHANGED"
  | "MODIFIED"
  | "ADDED"
  | "REMOVED"
  | "UNRESOLVED"
  | string;

export type RiskLevelValue =
  | "CRITICAL"
  | "HIGH"
  | "MEDIUM"
  | "LOW"
  | string;

export type VerificationStatusValue =
  | "VERIFIED"
  | "PARTIALLY_VERIFIED"
  | "UNVERIFIED"
  | "INSUFFICIENT_EVIDENCE"
  | "INVALID"
  | string;

export type ContractDocumentRef = {
  document_id?: string | null;
  document_version_id?: string | null;
  title?: string | null;
  workspace_id?: string | null;
};

export type ContractEvidenceRef = {
  evidence_id?: string | null;
  side?: string | null;
  document_id?: string | null;
  document_version_id?: string | null;
  clause_id?: string | null;
  identity_key?: string | null;
  chunk_id?: string | null;
  page_number?: number | null;
  start_offset?: number | null;
  end_offset?: number | null;
  source_type?: string | null;
  role?: string | null;
  display_text?: string | null;
};

export type ContractExactDifference = {
  change_type?: string | null;
  value_type?: string | null;
  direction?: string | null;
  old?: {
    raw?: string | null;
    value?: string | null;
    currency?: string | null;
    start?: number | null;
    end?: number | null;
  } | null;
  new?: {
    raw?: string | null;
    value?: string | null;
    currency?: string | null;
    start?: number | null;
    end?: number | null;
  } | null;
  delta?: string | null;
  relative_change_percent?: string | null;
  delta_unit?: string | null;
  context?: string | null;
  source_offset?: [number, number] | number[] | null;
  target_offset?: [number, number] | number[] | null;
  [key: string]: unknown;
};

export type ContractClauseRisk = {
  risk_category?: string | null;
  risk_level?: RiskLevelValue | null;
  risk_score?: string | number | null;
  risk_impact?: string | null;
  status?: string | null;
  triggered_rules?: string[] | null;
  reason?: string | null;
  [key: string]: unknown;
};

export type ContractClauseExplanation = {
  status?: string | null;
  reasons?: string[] | null;
  unavailable?: boolean;
  output?: {
    explanation?: string | null;
    evidence_ids?: string[] | null;
    [key: string]: unknown;
  } | null;
  [key: string]: unknown;
};

export type ContractEvidenceCheckStatus =
  | "VALID"
  | "INVALID"
  | "MISSING"
  | "MISMATCH"
  | "UNAVAILABLE"
  | string;

export type ContractEvidenceVerification = {
  evidence_id?: string | null;
  side?: string | null;
  status?: ContractEvidenceCheckStatus | null;
  reasons?: string[] | null;
};

export type ContractClauseVerification = {
  status?: VerificationStatusValue | null;
  human_message?: string | null;
  verified_evidence_ids?: string[] | null;
  invalid_evidence_ids?: string[] | null;
  evidence_results?: ContractEvidenceVerification[] | null;
  absence_status?: string | null;
  reasons?: string[] | null;
  [key: string]: unknown;
};

export type ContractClauseResult = {
  clause_id: string;
  v1_clause_id?: string | null;
  v2_clause_id?: string | null;
  status: ClauseComparisonStatus;
  mapping_confidence?: number | null;
  subtree_status?: string | null;
  exact_differences?: ContractExactDifference[];
  risk?: ContractClauseRisk | null;
  explanation?: ContractClauseExplanation | null;
  evidence?: ContractEvidenceRef[];
  citations?: ContractEvidenceRef[];
  verification?: ContractClauseVerification | null;
  v1_text?: string | null;
  v2_text?: string | null;
  finding_id?: string | null;
};

export type ContractComparisonSummary = {
  total_clauses: number;
  unchanged: number;
  modified: number;
  added: number;
  removed: number;
};

export type ContractComparisonStatistics = {
  total_clauses_compared?: number;
  unchanged?: number;
  modified?: number;
  added?: number;
  removed?: number;
  unresolved?: number;
  risk_counts?: Partial<Record<"critical" | "high" | "medium" | "low" | string, number>>;
  llm_calls?: number;
  llm_tokens?: number;
  processing_time_ms?: number;
  verification_rate?: number;
  citation_verification_rate?: number;
};

export type ContractComparisonMetadata = {
  comparison_id?: string | null;
  workspace_id?: string | null;
  document_v1?: ContractDocumentRef | null;
  document_v2?: ContractDocumentRef | null;
  created_at?: string | null;
  status?: string | null;
  quality_status?: string | null;
  quality_reasons?: string[] | null;
  explanation_incomplete?: boolean | null;
};

export type ContractComparisonReport = {
  metadata?: ContractComparisonMetadata | null;
  summary?: ContractComparisonSummary | null;
  statistics?: ContractComparisonStatistics | null;
  clauses?: {
    unchanged?: ContractClauseResult[];
    modified?: ContractClauseResult[];
    added?: ContractClauseResult[];
    removed?: ContractClauseResult[];
    unresolved?: ContractClauseResult[];
  } | null;
  risks?: unknown[];
  citations?: unknown[];
};

export type ComparisonResult = {
  similarities: string[];
  differences: string[];
  contract_comparison?: ContractComparisonReport | null;
};

export type ComparisonReviewStatus =
  | "OPEN"
  | "REVIEWED"
  | "NEEDS_ATTENTION"
  | "ACKNOWLEDGED"
  | string;

export type ComparisonReviewDecision = {
  status: ComparisonReviewStatus;
  reviewer_id?: string | null;
  reviewer_name?: string | null;
  reviewed_at?: string | null;
};

export type ComparisonCommentTarget =
  | "CLAUSE"
  | "FINDING"
  | "EXACT_DIFFERENCE"
  | "EVIDENCE"
  | string;

export type ComparisonComment = {
  id: string;
  clause_id: string;
  target_type: ComparisonCommentTarget;
  target_id?: string | null;
  body: string;
  author_id?: string | null;
  author_name?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
};

export type Comparison = {
  id: string;
  workspace_id: string;
  document_ids: string[];
  status: ComparisonStatus;
  result: ComparisonResult | null;
  review?: Record<string, ComparisonReviewDecision> | null;
  comments?: ComparisonComment[] | null;
  created_at: string;
};

export type ComparisonAuditAction =
  | "CLAUSE_OPENED"
  | "REVIEW_STATUS_CHANGED"
  | "COMMENT_ADDED"
  | "COMMENT_EDITED"
  | "COMMENT_DELETED"
  | string;

export type ComparisonAuditEvent = {
  id: string;
  action: ComparisonAuditAction;
  clause_id?: string | null;
  actor_id?: string | null;
  actor_name?: string | null;
  occurred_at: string;
  before?: Record<string, unknown> | null;
  after?: Record<string, unknown> | null;
  target_type?: string | null;
  target_id?: string | null;
  comment_id?: string | null;
};

export type ComparisonAuditTrail = {
  events: ComparisonAuditEvent[];
};

export type ComparisonCreateRequest = {
  document_ids: string[];
  focus?: string | null;
};

export type DocumentMeta = {
  title: string;
  created_at?: string | null;
};
